consensus/core/src/tx.rs

```diff
@@ -30,27 +35,57 @@ pub use smallvec::smallvec as scriptvec;
 pub type ScriptPublicKeys = HashSet<ScriptPublicKey>;
 
 /// Represents a Kaspad ScriptPublicKey
-#[derive(Default, Debug, PartialEq, Eq, Serialize, Clone, Hash)]
-#[serde(rename_all = "camelCase")]
+#[derive(Default, Debug, PartialEq, Eq, Clone, Hash)]
 pub struct ScriptPublicKey {
     version: ScriptPublicKeyVersion,
     script: ScriptVec, // Kept private to preserve read-only semantics
 }
 
+#[derive(Default, Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+#[serde(rename_all = "camelCase")]
+struct ScriptPublicKeyInternal<'a> {
+    version: ScriptPublicKeyVersion,
+    script: &'a [u8],
+}
+
+impl Serialize for ScriptPublicKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if serializer.is_human_readable() {
+            let mut hex = [0u8; SCRIPT_VECTOR_SIZE * 2 + 4];
+            faster_hex::hex_encode(&self.version.to_be_bytes(), &mut hex).expect("Unable to serialize ScriptPubKey version");
+            faster_hex::hex_encode(&self.script, &mut hex[4..]).expect("Unable to serialize ScriptPubKey script");
+            serializer.serialize_str(str::from_utf8(&hex).expect("hex is must be a valid UTF-8"))
+        } else {
+            ScriptPublicKeyInternal { version: self.version, script: &self.script }.serialize(serializer)
+        }
+    }
+}
+
 impl<'de: 'a, 'a> Deserialize<'de> for ScriptPublicKey {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: Deserializer<'de>,
     {
-        #[derive(Default, Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
-        #[serde(rename_all = "camelCase")]
-        struct ScriptPublicKeyInternal<'a> {
-            version: ScriptPublicKeyVersion,
-            script: &'a [u8],
+        if deserializer.is_human_readable() {
+            let s = <std::string::String as Deserialize>::deserialize(deserializer)?;
```

Comment:
```suggestion
            let s = <&str as Deserialize>::deserialize(deserializer)?;
```
It needs to be tested, I think it can prevent String allocation, but I'm not sure